lib/WooCommerce/Subscription.php

```diff
@@ -148,44 +148,59 @@ public function subscribeOnCheckout($orderId, $data) {
     $subscriber = Subscriber::where('email', $data['billing_email'])
       ->where('is_woocommerce_user', 1)
       ->findOne();
+
     if (!$subscriber) {
       // no subscriber: WooCommerce sync didn't work
       return null;
     }
 
-    $checkoutOptinEnabled = (bool)$this->settings->get(self::OPTIN_ENABLED_SETTING_NAME);
+    $checkoutOptinEnabled = (bool)$this->settings->get( self::OPTIN_ENABLED_SETTING_NAME );
+    $checkoutOptin = ! empty( $_POST[self::CHECKOUT_OPTIN_INPUT_NAME] );
+
+    return $this->handleSubscriberOptin( $subscriber, $checkoutOptinEnabled && $checkoutOptin );
+  }
+
+  /**
+   * Subscribe or unsubscribe a subscriber.
+   *
+   * @param Subscriber $subscriber Subscriber object
+   * @param bool $optin Opting in or opting out.
+   */
+  public function handleSubscriberOptin( Subscriber $subscriber, $optin = true ) {
```

Comment:
Needed access to this logic from the block initialisation class so I made it more general use.